File: 2_TimeSeriesErrorsAutocorrelation/TimeSeriesErrorsAutocorrelation.R
# ======================================================================
# TIME SERIES ERRORS AND AUTOCORRELATION ANALYSIS
# ======================================================================

# This program demonstrates:
# 1. Testing for autocorrelation and plotting autocorrelation functions (ACFs)
# 2. Heteroskedasticity-Consistent (HC) and Heteroskedasticity and Autocorrelation
#    Consistent (HAC) standard errors in regression models
# 3. Testing for serial correlation in regression residuals
# 4. Methods to adjust models for autocorrelation

# --------------------- SETUP AND PACKAGE LOADING ----------------------
# Load required packages
library(quantmod)    # For financial data retrieval
library(MTS)         # For multivariate time series analysis
library(forecast)    # For time series forecasting and visualization
library(sandwich)    # For robust covariance matrix estimators
library(lmtest)      # For testing linear regression models
library(car)         # For regression diagnostics and testing
library(dynlm)       # For dynamic linear models

# ---------------------- DATA ACQUISITION ------------------------------
# Model for determining price response of oil & gas drilling

# Read from FRED database:
# - WTI oil prices (MCOILWTICO)
# - Oil & gas drilling index (IPN213111N)
# - Henry Hub natural gas prices (MHHNGSP)
getSymbols("MCOILWTICO", src="FRED")   # WTI crude oil prices
getSymbols("IPN213111N", src="FRED")   # Oil & gas drilling index
getSymbols("MHHNGSP", src="FRED")      # Henry Hub natural gas prices

# ---------------------- DATA PREPARATION -----------------------------
# Merge all series into one data frame
price_drilling_levels = merge.xts(MHHNGSP, MCOILWTICO, IPN213111N, all=TRUE)

# Plot level data (appears non-stationary - will test formally later)
plot(price_drilling_levels, 
     main="Oil and Gas Prices and Drilling Activity (Levels)",
     col=c("blue", "red", "green"))
MTSplot(price_drilling_levels)

# Calculate log percentage changes (likely stationary)
price_drilling_changes = na.omit(diff(log(price_drilling_levels)))

# Plot transformed data
plot(price_drilling_changes, 
     main="Oil and Gas Prices and Drilling Activity (Log Changes)",
     col=c("blue", "red", "green"))
MTSplot(price_drilling_changes)

# ------------------- AUTOCORRELATION ANALYSIS ------------------------
# Investigate autocorrelation patterns in each series

# Function to test and visualize autocorrelation
analyze_autocorrelation <- function(series, name, max_lag=NULL) {
  # Determine appropriate lag order based on sample size
  if(is.null(max_lag)) {
    max_lag = round(log(length(series)))
  }
  
  # Plot ACF
  acf_plot <- acf(series, main=paste("ACF of", name), lag.max=max_lag)
  
  # Perform Box-Ljung test
  box_test_result <- Box.test(series, lag=max_lag, type='Ljung')
  
  # Print results
  cat("======================================\n")
  cat("Autocorrelation analysis for", name, "\n")
  cat("======================================\n")
  cat("Box-Ljung test (lag =", max_lag, "):\n")
  cat("Q statistic =", round(box_test_result$statistic, 4), "\n")
  cat("p-value =", round(box_test_result$p.value, 4), "\n")
  cat("--------------------------------------\n\n")
  
  return(list(acf=acf_plot, box_test=box_test_result))
}

# Determine appropriate lag order based on sample size
max_lag <- round(log(length(price_drilling_changes)))
cat("Suggested lag order based on log(T):", max_lag, "\n\n")

# Analyze each series
oil_autocorr <- analyze_autocorrelation(price_drilling_changes$MCOILWTICO, 
                                      "Oil Price Changes", max_lag)
gas_autocorr <- analyze_autocorrelation(price_drilling_changes$MHHNGSP, 
                                      "Natural Gas Price Changes", max_lag)
drilling_autocorr <- analyze_autocorrelation(price_drilling_changes$IPN213111N, 
                                          "Drilling Activity Changes", max_lag)

# -------------------- BASELINE REGRESSION MODEL ----------------------
# Estimate a simple model: How do prices affect drilling activity?

baseline_model <- lm(IPN213111N ~ MCOILWTICO + MHHNGSP, data=price_drilling_changes)
cat("Baseline regression model: How do prices affect drilling activity?\n")
cat("--------------------------------------------------------------\n")
print(summary(baseline_model))

# ----------------- RESIDUAL AUTOCORRELATION ANALYSIS -----------------
# Check for autocorrelation in the residuals

# Function to analyze residuals for autocorrelation
analyze_residuals <- function(model, df_correction=0, max_lag=NULL) {
  residuals <- residuals(model)
  
  # Determine appropriate lag order based on sample size if not provided
  if(is.null(max_lag)) {
    max_lag = round(log(length(residuals)))
  }
  
  # Visualize residuals
  par(mfrow=c(3,1))
  plot(residuals, type="l", main="Residuals Over Time", 
       xlab="Time", ylab="Residual")
  acf(residuals, main="ACF of Residuals", lag.max=max_lag)
  pacf(residuals, main="PACF of Residuals", lag.max=max_lag)
  par(mfrow=c(1,1))
  
  # Perform Box-Ljung test with degrees of freedom correction
  box_test_result <- Box.test(residuals, lag=max_lag, type='Ljung')
  df <- max_lag - df_correction
  adjusted_p_value <- 1-pchisq(box_test_result$statistic, df)
  
  # Print results
  cat("====================================================\n")
  cat("Residual autocorrelation analysis\n")
  cat("====================================================\n")
  cat("Box-Ljung test (lag =", max_lag, "):\n")
  cat("Q statistic =", round(box_test_result$statistic, 4), "\n")
  cat("Unadjusted p-value =", round(box_test_result$p.value, 4), "\n")
  cat("Degrees of freedom correction =", df_correction, "\n")
  cat("Adjusted degrees of freedom =", df, "\n")
  cat("Adjusted p-value =", round(adjusted_p_value, 4), "\n")
  cat("----------------------------------------------------\n\n")
  
  return(list(
    residuals=residuals, 
    box_test=box_test_result, 
    adjusted_p_value=adjusted_p_value
  ))
}

# Analyze residuals from baseline model
# DF correction = 2 (for 2 parameters estimated in model)
residual_analysis <- analyze_residuals(baseline_model, df_correction=2)

# Also use the tsdisplay function for a comprehensive view
tsdisplay(residuals(baseline_model), main="Comprehensive Residual Analysis")

# ----------------- HC AND HAC STANDARD ERRORS -----------------------
# Apply robust standard errors to account for issues in the error terms

# Function to display different standard error calculations
compare_standard_errors <- function(model) {
  # Original standard errors
  original <- coeftest(model)
  
  # HAC standard errors
  hac <- coeftest(model, vcov=vcovHAC(model))
  
  # HC standard errors
  hc <- coeftest(model, vcov=vcovHC(model))
  
  # Create comparison table
  coef_names <- rownames(original)
  result_table <- data.frame(
    Coefficient = coef_names,
    Estimate = original[,1],
    Original_SE = original[,2],
    Original_p = original[,4],
    HC_SE = hc[,2],
    HC_p = hc[,4],
    HAC_SE = hac[,2],
    HAC_p = hac[,4]
  )
  
  return(result_table)
}

# Compare standard errors for baseline model
std_error_comparison <- compare_standard_errors(baseline_model)
cat("Comparison of Standard Error Methods\n")
cat("====================================\n")
print(std_error_comparison)

# F-test with HAC standard errors
gas_price_test <- linearHypothesis(baseline_model, c("MHHNGSP=0"), 
                                 vcov=vcovHAC(baseline_model), 
                                 test="F")
cat("\nF-test for Natural Gas Price coefficient (HAC std errors)\n")
cat("-------------------------------------------------------\n")
print(gas_price_test)

# Calculate corresponding t-statistic
t_statistic <- sqrt(gas_price_test$F[2])
cat("Corresponding t-statistic:", round(t_statistic, 4), "\n\n")

# ------------------- ADJUSTING FOR AUTOCORRELATION -----------------
# Two approaches to handle autocorrelation:
# 1. Modeling the residuals (ARIMA)
# 2. Including lags of dependent/independent variables (dynlm)

# Convert data to ts object for time series modeling
price_drilling_ts <- ts(price_drilling_changes, freq=12, start=c(1997, 2))

# ---------------- APPROACH 1: ARIMA MODEL --------------------------
# Model the residuals using ARIMA

# Variables are:
# Column 1: MHHNGSP (Natural Gas Price)
# Column 2: MCOILWTICO (Oil Price)
# Column 3: IPN213111N (Drilling Activity)

# Estimate ARIMA model with AR(2) for residuals
arima_model <- arima(price_drilling_changes[,3], order=c(2,0,0),
                   xreg=cbind(price_drilling_changes[,1], price_drilling_changes[,2]),
                   include.mean=TRUE)

cat("ARIMA Model Results (AR(2) for residuals)\n")
cat("----------------------------------------\n")
print(summary(arima_model))

# Check if autocorrelation has been resolved
arima_residual_analysis <- analyze_residuals(arima_model, df_correction=4)

# Standard errors from ARIMA model
arima_coef_test <- coeftest(arima_model, vcov=vcov(arima_model))
cat("ARIMA model coefficients with standard errors:\n")
print(arima_coef_test)

# -------------- APPROACH 2: DYNAMIC LINEAR MODEL ------------------
# Include lags of variables to account for autocorrelation

# Model with lags of the dependent variable (drilling activity)
dynamic_model <- dynlm(price_drilling_ts[,"IPN213111N"] ~ 
                     L(price_drilling_ts[,"IPN213111N"], c(1:2)) +
                     price_drilling_ts[,"MHHNGSP"] + 
                     price_drilling_ts[,"MCOILWTICO"])

cat("Dynamic Linear Model Results (with lags of drilling)\n")
cat("--------------------------------------------------\n")
print(summary(dynamic_model))

# Check residuals of dynamic model
tsdisplay(residuals(dynamic_model), 
         main="Residuals of Dynamic Model with Drilling Lags")

# Apply HC standard errors to dynamic model
dynamic_hc_test <- coeftest(dynamic_model, vcov=vcovHC(dynamic_model))
cat("Dynamic model coefficients with HC standard errors:\n")
print(dynamic_hc_test)

# Test correlation between oil and gas prices
price_correlation <- cor(price_drilling_changes$MHHNGSP, 
                       price_drilling_changes$MCOILWTICO)
cat("Correlation between oil and gas prices:", round(price_correlation, 4), "\n\n")

# Joint test of price effects using HC standard errors
# Reestimate with simpler variable names for easier joint testing
dynamic_model_simple <- dynlm(price_drilling_ts[,3] ~ 
                           L(price_drilling_ts[,3], c(1:2)) +
                           price_drilling_ts[,1] + 
                           price_drilling_ts[,2])

joint_price_test <- linearHypothesis(dynamic_model_simple, 
                                   c("price_drilling_ts[, 1]=0", 
                                     "price_drilling_ts[, 2]=0"),
                                   vcov=vcovHC(dynamic_model_simple), 
                                   test="F")
cat("Joint test of price effects (HC std errors):\n")
print(joint_price_test)

# -------------- EXTENDED DYNAMIC MODEL WITH PRICE LAGS -------------
# Try including lags of price variables

extended_dynamic_model <- dynlm(price_drilling_ts[,3] ~ 
                             L(price_drilling_ts[,3], c(1:2)) + 
                             L(price_drilling_ts[,1], c(1)) +
                             L(price_drilling_ts[,2], c(0:4)))

cat("Extended Dynamic Model Results (with price lags)\n")
cat("-----------------------------------------------\n")
print(summary(extended_dynamic_model))

# Check residuals
tsdisplay(residuals(extended_dynamic_model), 
         main="Residuals of Extended Dynamic Model with Price Lags")

# Apply HAC standard errors
extended_hac_test <- coeftest(extended_dynamic_model, vcov=vcovHAC(extended_dynamic_model))
cat("Extended dynamic model coefficients with HAC standard errors:\n")
print(extended_hac_test)

# ---------------------- SUMMARY OF FINDINGS ------------------------
cat("\n===========================================================\n")
cat("SUMMARY OF FINDINGS\n")
cat("===========================================================\n")
cat("1. Baseline model shows significant autocorrelation in residuals\n")
cat("2. HAC standard errors correct for this but don't address underlying issue\n")
cat("3. ARIMA model successfully accounts for residual autocorrelation\n")
cat("4. Dynamic model with lags of drilling also addresses autocorrelation\n")
cat("5. Extended model suggests lagged price effects might be important\n")
cat("6. Joint test of contemporaneous price effects is not significant\n")
cat("-----------------------------------------------------------\n")

File: 2_TimeSeriesErrorsAutocorrelation/python_versions/TimeSeriesErrorsAutocorrelation.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# =============================================================================
# TIME SERIES ERRORS AND AUTOCORRELATION ANALYSIS
# =============================================================================

# This program demonstrates:
# 1. Testing for autocorrelation and plotting autocorrelation functions (ACFs)
# 2. Heteroskedasticity-Consistent (HC) and Heteroskedasticity and Autocorrelation
#    Consistent (HAC) standard errors in regression models
# 3. Testing for serial correlation in regression residuals
# 4. Methods to adjust models for autocorrelation

# Import required packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from pandas_datareader import data as pdr
import statsmodels.api as sm
from statsmodels.graphics.tsaplots import plot_acf, plot_pacf
from statsmodels.stats.diagnostic import acorr_ljungbox
from statsmodels.tsa.arima.model import ARIMA
from statsmodels.tsa.stattools import acf, pacf
from statsmodels.regression.linear_model import OLS
from statsmodels.stats.sandwich_covariance import cov_hac, cov_white_simple
from statsmodels.stats.diagnostic import het_white
from statsmodels.iolib.summary2 import summary_col
import yfinance as yf
from scipy import stats
import warnings
from datetime import datetime

# Set some display options
plt.style.use('seaborn-v0_8-whitegrid')
plt.rcParams['figure.figsize'] = (12, 8)
plt.rcParams['font.size'] = 12
warnings.filterwarnings("ignore")

# Use yfinance for data retrieval
#yf.pdr_override()

# =============================================================================
# DATA ACQUISITION AND PREPARATION
# =============================================================================

def fetch_fred_data(series_ids, start_date='1997-01-01', end_date=None):
    """
    Fetch data from FRED using pandas_datareader
    
    Parameters:
    -----------
    series_ids : list
        List of FRED series IDs
    start_date : str
        Start date in YYYY-MM-DD format
    end_date : str
        End date in YYYY-MM-DD format (defaults to today)
        
    Returns:
    --------
    pd.DataFrame
        DataFrame containing requested series
    """
    if end_date is None:
        end_date = datetime.now().strftime('%Y-%m-%d')
    
    # Initialize empty DataFrame
    data = pd.DataFrame()
    
    # Fetch each series
    for series_id in series_ids:
        try:
            series = pdr.get_data_fred(series_id, start_date, end_date)
            data = pd.concat([data, series], axis=1)
        except Exception as e:
            print(f"Error fetching {series_id}: {e}")
    
    return data

# Fetch data from FRED
# - WTI oil prices (MCOILWTICO)
# - Oil & gas drilling index (IPN213111N)
# - Henry Hub natural gas prices (MHHNGSP)
series_ids = ['MCOILWTICO', 'IPN213111N', 'MHHNGSP']
price_drilling_levels = fetch_fred_data(series_ids)

# Reorder columns to match the R version (Gas, Oil, Drilling)
price_drilling_levels = price_drilling_levels[['MHHNGSP', 'MCOILWTICO', 'IPN213111N']]

# Plot level data (appears non-stationary)
fig, ax = plt.subplots(figsize=(14, 8))
price_drilling_levels.plot(ax=ax)
ax.set_title('Oil and Gas Prices and Drilling Activity (Levels)')
ax.set_xlabel('Date')
ax.set_ylabel('Value')
plt.legend(['Henry Hub Natural Gas Price', 'WTI Crude Oil Price', 'Oil & Gas Drilling Index'])
plt.tight_layout()
plt.show()

# Calculate log percentage changes (likely stationary)
price_drilling_changes = np.log(price_drilling_levels).diff().dropna() * 100

# Plot transformed data
fig, ax = plt.subplots(figsize=(14, 8))
price_drilling_changes.plot(ax=ax)
ax.set_title('Oil and Gas Prices and Drilling Activity (Log Changes)')
ax.set_xlabel('Date')
ax.set_ylabel('Percentage Change')
plt.legend(['Henry Hub Natural Gas Price', 'WTI Crude Oil Price', 'Oil & Gas Drilling Index'])
plt.tight_layout()
plt.show()

# =============================================================================
# AUTOCORRELATION ANALYSIS
# =============================================================================

def analyze_autocorrelation(series, name, max_lag=None):
    """
    Test and visualize autocorrelation in a time series
    
    Parameters:
    -----------
    series : array-like
        Time series data
    name : str
        Name of the series for plot titles
    max_lag : int or None
        Maximum lag to consider (defaults to log of sample size)
    
    Returns:
    --------
    dict
        Dictionary containing ACF, PACF and Ljung-Box test results
    """
    # Handle NaN values
    series = series.dropna()
    
    # Determine appropriate lag order based on sample size
    if max_lag is None:
        max_lag = int(np.round(np.log(len(series))))
    
    # Plot ACF and PACF
    fig, axes = plt.subplots(2, 1, figsize=(12, 10))
    
    plot_acf(series, lags=max_lag, ax=axes[0])
    axes[0].set_title(f'Autocorrelation Function (ACF) of {name}')
    
    plot_pacf(series, lags=max_lag, ax=axes[1])
    axes[1].set_title(f'Partial Autocorrelation Function (PACF) of {name}')
    
    plt.tight_layout()
    plt.show()
    
    # Perform Ljung-Box test
    lb_test = acorr_ljungbox(series, lags=max_lag, return_df=True)
    
    # Print results
    print("="*50)
    print(f"Autocorrelation analysis for {name}")
    print("="*50)
    print(f"Ljung-Box test (lag = {max_lag}):")
    print(f"Q statistic = {lb_test['lb_stat'].iloc[-1]:.4f}")
    print(f"p-value = {lb_test['lb_pvalue'].iloc[-1]:.4f}")
    print("-"*50, "\n")
    
    return {
        'series': series,
        'ljung_box_test': lb_test,
        'acf': acf(series, nlags=max_lag),
        'pacf': pacf(series, nlags=max_lag)
    }

# Determine appropriate lag order based on sample size
max_lag = int(np.round(np.log(len(price_drilling_changes))))
print(f"Suggested lag order based on log(T): {max_lag}\n")

# Analyze each series
oil_autocorr = analyze_autocorrelation(price_drilling_changes['MCOILWTICO'], 
                                     "Oil Price Changes", max_lag)
gas_autocorr = analyze_autocorrelation(price_drilling_changes['MHHNGSP'], 
                                     "Natural Gas Price Changes", max_lag)
drilling_autocorr = analyze_autocorrelation(price_drilling_changes['IPN213111N'], 
                                         "Drilling Activity Changes", max_lag)

# =============================================================================
# BASELINE REGRESSION MODEL
# =============================================================================

def run_regression(y, X, add_constant=True):
    """
    Run OLS regression and return results
    
    Parameters:
    -----------
    y : array-like
        Dependent variable
    X : array-like
        Independent variables
    add_constant : bool
        Whether to add a constant term (intercept)
    
    Returns:
    --------
    statsmodels.regression.linear_model.RegressionResults
        Regression results
    """
    # Add constant if requested
    if add_constant:
        X = sm.add_constant(X)
    
    # Fit model
    model = sm.OLS(y, X)
    results = model.fit()
    
    return results

# Prepare data for baseline model
Y = price_drilling_changes['IPN213111N']
X = price_drilling_changes[['MCOILWTICO', 'MHHNGSP']]

# Estimate baseline model
baseline_model = run_regression(Y, X)

# Display results
print("Baseline regression model: How do prices affect drilling activity?")
print("="*60)
print(baseline_model.summary())

# =============================================================================
# RESIDUAL AUTOCORRELATION ANALYSIS
# =============================================================================

def analyze_residuals(model_results, df_correction=0, max_lag=None):
    """
    Analyze residuals for autocorrelation
    
    Parameters:
    -----------
    model_results : statsmodels.regression.linear_model.RegressionResults
        Results from regression model
    df_correction : int
        Degrees of freedom correction for Ljung-Box test
    max_lag : int or None
        Maximum lag to consider (defaults to log of sample size)
    
    Returns:
    --------
    dict
        Dictionary containing residual analysis results
    """
    residuals = model_results.resid
    
    # Determine appropriate lag order based on sample size if not provided
    if max_lag is None:
        max_lag = int(np.round(np.log(len(residuals))))
    
    # Create plots
    fig, axes = plt.subplots(3, 1, figsize=(12, 12))
    
    # Time series plot of residuals
    axes[0].plot(residuals.index, residuals)
    axes[0].set_title('Residuals Over Time')
    axes[0].set_xlabel('Time')
    axes[0].set_ylabel('Residual')
    
    # ACF plot
    plot_acf(residuals, lags=max_lag, ax=axes[1])
    axes[1].set_title('ACF of Residuals')
    
    # PACF plot
    plot_pacf(residuals, lags=max_lag, ax=axes[2])
    axes[2].set_title('PACF of Residuals')
    
    plt.tight_layout()
    plt.show()
    
    # Perform Ljung-Box test
    lb_test = acorr_ljungbox(residuals, lags=max_lag, return_df=True)
    
    # Adjust degrees of freedom
    df = max_lag - df_correction
    adjusted_pvalue = 1 - stats.chi2.cdf(lb_test['lb_stat'].iloc[-1], df)
    
    # Print results
    print("="*60)
    print("Residual autocorrelation analysis")
    print("="*60)
    print(f"Ljung-Box test (lag = {max_lag}):")
    print(f"Q statistic = {lb_test['lb_stat'].iloc[-1]:.4f}")
    print(f"Unadjusted p-value = {lb_test['lb_pvalue'].iloc[-1]:.4f}")
    print(f"Degrees of freedom correction = {df_correction}")
    print(f"Adjusted degrees of freedom = {df}")
    print(f"Adjusted p-value = {adjusted_pvalue:.4f}")
    print("-"*60, "\n")
    
    return {
        'residuals': residuals,
        'ljung_box_test': lb_test,
        'adjusted_pvalue': adjusted_pvalue,
        'acf': acf(residuals, nlags=max_lag),
        'pacf': pacf(residuals, nlags=max_lag)
    }

# Analyze residuals from baseline model (df_correction = 2 for 2 parameters)
residual_analysis = analyze_residuals(baseline_model, df_correction=2)

# =============================================================================
# HC AND HAC STANDARD ERRORS
# =============================================================================

def compare_standard_errors(model_results):
    """
    Calculate and compare different standard error estimates
    
    Parameters:
    -----------
    model_results : statsmodels.regression.linear_model.RegressionResults
        Results from regression model
    
    Returns:
    --------
    pd.DataFrame
        DataFrame comparing different standard error estimates
    """
    # Original standard errors
    original_cov = model_results.cov_params()
    original_bse = model_results.bse
    
    # HAC standard errors
    hac_cov = cov_hac(model_results)
    hac_bse = np.sqrt(np.diag(hac_cov))
    
    # HC standard errors (White's heteroskedasticity-consistent)
    hc_cov = cov_white_simple(model_results)
    hc_bse = np.sqrt(np.diag(hc_cov))
    
    # Create t-statistics and p-values
    params = model_results.params
    
    # Original
    original_tvalues = params / original_bse
    original_pvalues = 2 * (1 - stats.t.cdf(np.abs(original_tvalues), model_results.df_resid))
    
    # HC
    hc_tvalues = params / hc_bse
    hc_pvalues = 2 * (1 - stats.t.cdf(np.abs(hc_tvalues), model_results.df_resid))
    
    # HAC
    hac_tvalues = params / hac_bse
    hac_pvalues = 2 * (1 - stats.t.cdf(np.abs(hac_tvalues), model_results.df_resid))
    
    # Create comparison table
    results = pd.DataFrame({
        'Coefficient': params.index,
        'Estimate': params.values,
        'Original_SE': original_bse,
        'Original_t': original_tvalues,
        'Original_p': original_pvalues,
        'HC_SE': hc_bse,
        'HC_t': hc_tvalues,
        'HC_p': hc_pvalues,
        'HAC_SE': hac_bse,
        'HAC_t': hac_tvalues,
        'HAC_p': hac_pvalues
    })
    
    print("Comparison of Standard Error Methods")
    print("="*50)
    print(results[['Coefficient', 'Estimate', 'Original_SE', 'Original_p', 'HC_SE', 'HC_p', 'HAC_SE', 'HAC_p']])
    
    return results

# Compare standard errors for baseline model
std_error_comparison = compare_standard_errors(baseline_model)

# =============================================================================
# ADJUSTING FOR AUTOCORRELATION - APPROACH 1: ARIMA MODEL
# =============================================================================

def fit_arima_with_regressors(endog, exog, order=(2,0,0), include_mean=True):
    """
    Fit ARIMA model with external regressors
    
    Parameters:
    -----------
    endog : array-like
        Dependent variable (time series)
    exog : array-like
        Exogenous variables (regressors)
    order : tuple
        ARIMA order specification (p, d, q)
    include_mean : bool
        Whether to include a mean term in the model
    
    Returns:
    --------
    statsmodels.tsa.arima.model.ARIMAResults
        ARIMA model results
    """
    # Add constant if requested
    if include_mean:
        exog = sm.add_constant(exog)
    
    # Fit ARIMA model
    model = ARIMA(endog, exog=exog, order=order, trend='n')
    results = model.fit()
    
    return results

# Prepare data for ARIMA model
# Variables are:
# Column 0: MHHNGSP (Natural Gas Price)
# Column 1: MCOILWTICO (Oil Price)
# Column 2: IPN213111N (Drilling Activity)

# Fit ARIMA(2,0,0) model with regressors
arima_model = fit_arima_with_regressors(
    price_drilling_changes['IPN213111N'],
    price_drilling_changes[['MHHNGSP', 'MCOILWTICO']]
)

# Display results
print("ARIMA Model Results (AR(2) for residuals)")
print("="*50)
print(arima_model.summary())

# Analyze residuals to check if autocorrelation has been resolved
arima_residuals = arima_model.resid
analyze_autocorrelation(arima_residuals, "ARIMA Model Residuals")

# =============================================================================
# ADJUSTING FOR AUTOCORRELATION - APPROACH 2: DYNAMIC MODEL
# =============================================================================

def fit_dynamic_model(y, X, y_lags=None, X_lags=None):
    """
    Fit dynamic regression model with lags
    
    Parameters:
    -----------
    y : pd.Series
        Dependent variable
    X : pd.DataFrame
        Independent variables
    y_lags : list or None
        List of lags for dependent variable
    X_lags : dict or None
        Dictionary of lags for independent variables {column_name: [lags]}
    
    Returns:
    --------
    statsmodels.regression.linear_model.RegressionResults
        Regression results
    """
    # Create copy of data to avoid modifying original
    data = pd.concat([y, X], axis=1)
    
    # Create lagged dependent variables
    if y_lags is not None:
        for lag in y_lags:
            data[f'{y.name}_lag{lag}'] = y.shift(lag)
    
    # Create lagged independent variables
    if X_lags is not None:
        for col, lags in X_lags.items():
            for lag in lags:
                data[f'{col}_lag{lag}'] = X[col].shift(lag)
    
    # Drop rows with NaN from lagged variables
    data = data.dropna()
    
    # Prepare variables for regression
    y_dynamic = data[y.name]
    
    # Create list of X variables (including lagged variables)
    X_columns = []
    if y_lags is not None:
        X_columns.extend([f'{y.name}_lag{lag}' for lag in y_lags])
    
    # Add original X variables
    X_columns.extend(X.columns)
    
    # Add lagged X variables
    if X_lags is not None:
        for col, lags in X_lags.items():
            for lag in lags:
                if lag > 0:  # Only add lagged, not contemporaneous
                    X_columns.append(f'{col}_lag{lag}')
    
    # Run regression
    X_dynamic = data[X_columns]
    model = run_regression(y_dynamic, X_dynamic)
    
    return model, data

# Model with lags of the dependent variable (drilling activity)
dynamic_model, dynamic_data = fit_dynamic_model(
    price_drilling_changes['IPN213111N'],
    price_drilling_changes[['MHHNGSP', 'MCOILWTICO']],
    y_lags=[1, 2]
)

# Display results
print("Dynamic Linear Model Results (with lags of drilling)")
print("="*60)
print(dynamic_model.summary())

# Analyze residuals to check if autocorrelation has been resolved
dynamic_residual_analysis = analyze_residuals(dynamic_model)

# Calculate HAC and HC standard errors for dynamic model
dynamic_se_comparison = compare_standard_errors(dynamic_model)

# Test correlation between oil and gas prices
price_correlation = price_drilling_changes['MHHNGSP'].corr(price_drilling_changes['MCOILWTICO'])
print(f"Correlation between oil and gas prices: {price_correlation:.4f}\n")

# =============================================================================
# EXTENDED DYNAMIC MODEL WITH PRICE LAGS
# =============================================================================

# Create extended model with lags of prices
extended_model, extended_data = fit_dynamic_model(
    price_drilling_changes['IPN213111N'],
    price_drilling_changes[['MHHNGSP', 'MCOILWTICO']],
    y_lags=[1, 2],
    X_lags={
        'MHHNGSP': [1],
        'MCOILWTICO': [0, 1, 2, 3, 4]
    }
)

# Display results
print("Extended Dynamic Model Results (with price lags)")
print("="*60)
print(extended_model.summary())

# Analyze residuals
extended_residual_analysis = analyze_residuals(extended_model)

# Calculate HAC standard errors for extended model
extended_hac_se = compare_standard_errors(extended_model)

# =============================================================================
# JOINT TESTS AND MODEL COMPARISON
# =============================================================================

def wald_test(model_results, restrictions, cov_type='nonrobust'):
    """
    Perform Wald test for joint hypothesis
    
    Parameters:
    -----------
    model_results : statsmodels.regression.linear_model.RegressionResults
        Results from regression model
    restrictions : list or ndarray
        List of restrictions to test (as a constraint matrix)
    cov_type : str
        Type of covariance estimator to use
    
    Returns:
    --------
    tuple
        (F-statistic, p-value, df1, df2)
    """
    # Perform Wald test
    wald_result = model_results.wald_test(restrictions, cov_p=cov_type)
    
    # Get F-statistic and p-value
    f_stat = wald_result.fvalue
    p_value = wald_result.pvalue
    df1 = wald_result.df_num
    df2 = wald_result.df_denom
    
    # Print results
    print(f"Wald test for joint hypothesis")
    print(f"F-statistic: {f_stat:.4f}")
    print(f"p-value: {p_value:.4f}")
    print(f"Degrees of freedom: ({df1}, {df2})")
    
    return f_stat, p_value, df1, df2

# Compare models (baseline, dynamic, and extended)
model_names = ['Baseline', 'Dynamic (with drilling lags)', 'Extended (with price lags)']
models = [baseline_model, dynamic_model, extended_model]

# Create comparison summary
model_comparison = summary_col(
    models, 
    model_names=model_names,
    stars=True,
    info_dict={
        'R-squared': lambda x: f"{x.rsquared:.4f}",
        'Adj. R-squared': lambda x: f"{x.rsquared_adj:.4f}",
        'No. observations': lambda x: f"{int(x.nobs)}",
        'Residual Std. Error': lambda x: f"{np.sqrt(x.scale):.4f}"
    }
)

print("\nModel Comparison")
print("="*100)
print(model_comparison)

# =============================================================================
# SUMMARY OF FINDINGS
# =============================================================================

print("\n" + "="*70)
print("SUMMARY OF FINDINGS")
print("="*70)
print("1. Baseline model shows significant autocorrelation in residuals")
print("2. HAC standard errors correct for this but don't address underlying issue")
print("3. ARIMA model successfully accounts for residual autocorrelation")
print("4. Dynamic model with lags of drilling also addresses autocorrelation")
print("5. Extended model suggests lagged price effects might be important")
print("6. Joint test of contemporaneous price effects is not significant")
print("-"*70)